src/managers/ApplicationCommandPermissionsManager.js

```diff
@@ -0,0 +1,410 @@
+        resolvedIDs.push(userID);
+      });
+    } else if (users) {
+      const userID = this.client.users.resolveID(users);
+      if (!userID) {
+        throw new TypeError('INVALID_TYPE', 'users', 'Array or UserResolvable');
+      }
+      resolvedIDs.push(userID);
+    }
+
+    if (Array.isArray(roles)) {
+      roles.forEach(role => {
+        if (typeof role === 'string') {
+          resolvedIDs.push(role);
+          return;
+        }
+        if (!this.guild) throw new Error('GUILD_UNCACHED_ROLE_RESOLVE');
+        const roleID = this.guild.roles.resolveID(role);
+        if (!roleID) throw new TypeError('INVALID_ELEMENT', 'Array', 'users', role);
+        resolvedIDs.push(roleID);
+      });
+    } else if (roles) {
+      if (typeof roles === 'string') {
+        resolvedIDs.push(roles);
+      } else {
+        if (!this.guild) throw new Error('GUILD_UNCACHED_ROLE_RESOLVE');
+        const roleID = this.guild.roles.resolveID(roles);
+        if (!roleID) {
+          throw new TypeError('INVALID_TYPE', 'users', 'Array or RoleResolvable');
+        }
+        resolvedIDs.push(roleID);
+      }
+    }
+    const permissions = existing.filter(perm => !resolvedIDs.includes(perm.id));
+
+    return this.set({ guild: guildID, command: commandID, permissions });
+  }
+
+  /**
+   * Options used to check existance of permissions on a command
+   * <warn>The `command` parameter is not optional when the managers `commandID` is `null`</warn>
+   * @typedef {BaseApplicationCommandPermissionsOptions} HasApplicationCommandPermissionsOptions
+   * @param {UserResolvable|RoleResolvable} permissionID The user or role to check if a permission exists for
+   * on this command.
+   */
+
+  /**
+   * Check whether a permission exists for a user or role
+   * @param {AddApplicationCommandPermissionsOptions} options Options used to check permissions
+   * @returns {Promise<boolean>}
+   * @example
+   * // Check whether a user has permission to use a command
+   * guild.commands.permissions.has({ command: '123456789012345678', permissionID: '876543210123456789' })
+   *  .then(console.log)
+   *  .catch(console.error);
+   */
+  async has({ guild, command, permissionID }) {
+    const { guildID, commandID } = this._validateOptions(guild, command);
+    if (!commandID) throw new TypeError('INVALID_TYPE', 'command', 'ApplicationCommandResolvable');
+
+    let existing = [];
+    try {
+      existing = await this.fetch({ guild: guildID, command: commandID });
+    } catch (error) {
+      if (error.code !== APIErrors.UNKNOWN_APPLICATION_COMMAND_PERMISSIONS) throw error;
+    }
+
+    if (!permissionID) throw new TypeError('INVALID_TYPE', 'permissionsID', 'UserResolvable or RoleResolvable');
+    let resolvedID = permissionID;
+    if (typeof permissionID !== 'string') {
+      resolvedID = this.client.users.resolveID(permissionID);
+      if (!resolvedID) {
+        if (!this.guild) throw new Error('GUILD_UNCACHED_ROLE_RESOLVE');
```

Comment:
Isn't this error needed across the board? Since permissions are guild-only

Reply:
No, *theoretically* it should be impossible to hit this error as you shouldn't be able to pass a Role object from an uncached guild unless you manually construct it I guess.

The missing guild / guild id is  handled in `_validateOptions`, the first thing it does, and that's the first thing called in every function.